File: indico/legacy/services/interface/rpc/json.py
# This file is part of Indico.
# Copyright (C) 2002 - 2021 CERN
#
# Indico is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see the
# LICENSE file for more details.

from flask import jsonify, request

from indico.core.logger import Logger
from indico.legacy.services.interface.rpc.process import invoke_method
from indico.util.fossilize import clearCache


def process():
    clearCache()
    payload = request.json
    Logger.get('rpc').info('json rpc request. request: %r', payload)
    rv = {}
    if 'id' in payload:
        rv['id'] = payload['id']
    rv['result'] = invoke_method(str(payload['method']), payload.get('params', []))
    return jsonify(rv)

File: indico/web/client/js/legacy/libs/presentation/Data/Json.js
// This file is part of Indico.
// Copyright (C) 2002 - 2025 CERN
//
// Indico is free software; you can redistribute it and/or
// modify it under the terms of the MIT License; see the
// LICENSE file for more details.

var Json = {
  /**
   * Encodes the object into a json.
   * @param {Object} object
   * @return {String} json
   */
  write: function(object) {
    if (!exists(object)) {
      return 'null';
    }
    switch (typeof object) {
      case 'boolean':
      case 'number':
        return str(object);
      case 'string':
        return escapeString(object);
      case 'object':
        if (object.Getter) {
          return Json.write(object.get());
        }
        if (object.Dictionary) {
          object = object.getAll();
        }
        if (object.Enumerable) {
          return '[' + object.each(stacker(Json.write)).join(',') + ']';
        }
        if (isArray(object)) {
          return '[' + iterate(object, stacker(Json.write)).join(',') + ']';
        }
        var properties = [];
        enumerate(object, function(value, key) {
          if (!isFunction(value)) {
            properties.push(escapeString(key) + ':' + Json.write(value));
          }
        });
        return '{' + properties.join(',') + '}';
      default:
        throw 'Invalid object: ' + str(object);
    }
  },
};
